google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java

```diff
@@ -174,4 +162,27 @@ public void invalidate() {
   // No-op method needed to implement SessionTransaction interface.
   @Override
   public void setSpan(Span span) {}
+
+  private ByteString initTransaction() {
+    final BeginTransactionRequest request =
+        BeginTransactionRequest.newBuilder()
+            .setSession(session.getName())
+            .setOptions(
+                TransactionOptions.newBuilder()
+                    .setPartitionedDml(TransactionOptions.PartitionedDml.getDefaultInstance()))
+            .build();
+    Transaction txn = rpc.beginTransaction(request, session.getOptions());
+    if (txn.getId().isEmpty()) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INTERNAL,
+          "Failed to init transaction, missing transaction id\n" + session.getName());
+    }
+    return txn.getId();
+  }
+
+  private boolean shouldResumeOrRestartTransaction(Exception e) {
+    return e instanceof UnavailableException
+        || (e instanceof InternalException
+            && e.getMessage().contains("Received unexpected EOS on DATA frame from server"));
```

Comment:
This is hacky, but unfortunately we do not get a specific exception for this error, so we have to proxy through the exception message.

Reply:
The same error is also retried for streaming queries, and this check could probably use the existing check in `SpannerExceptionFactory`: https://github.com/googleapis/java-spanner/blob/b2641007c3ef77765ab8d98fba02bc6039321c46/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SpannerExceptionFactory.java#L233. It's still just as hacky, but it would keep it confined to one method.